### ISSUE TITLE
Add option to make end date of allDay events 'inclusive'

### ISSUE THREAD

**Issue by jnachtigall** [fullcalendar/fullcalendar#3190]
I know that the end date of `allDay` events is `exclusive` (http://fullcalendar.io/docs/event_data/Event_Object/). I know that this is _technically_ the same for ical. 

We have a backend/CMS where users enter the events and their dates. For example, a 3 days event from 16th June to 18th June. There is no way that I can teach my users that this event should be entered as 16th June to 19th June instead of what it actually is. Maybe this is also a cultural thing (germany in my case).

An option like `endIsInclusive` (default: false) would be very much appreciated. 

I found various requests for this issue: https://github.com/fullcalendar/fullcalendar/issues/3098, https://github.com/fullcalendar/fullcalendar/issues/270, https://github.com/fullcalendar/fullcalendar/issues/2653, https://github.com/fullcalendar/fullcalendar/issues/2517, https://github.com/fullcalendar/fullcalendar/issues/2909, https://github.com/fullcalendar/fullcalendar/issues/1028, https://github.com/fullcalendar/fullcalendar/issues/2640, https://github.com/fullcalendar/fullcalendar/issues/2268, https://github.com/fullcalendar/fullcalendar/issues/2344, http://stackoverflow.com/questions/25941499/fullcalendar-multi-day-event-spans-1-day-too-short#comment40626092_25942710

What makes this worse, there seems to be no easy workaround for this issue, see https://github.com/fullcalendar/fullcalendar/issues/2344#issuecomment-132572231

**Comment by atilleul** [comment]
I'm also very interested in this feature. In fact a number of our users don't understand why they need to set the next day to have 1 day: 2016-06-01 to 2016-06-01 is actually one day in a user-friendly interface, not: 2016-06-01 to 2016-06-02.
And showing in our edit pages "2016-06-01 00:00 to 2016-06-02 00:00" is not very nice.
Bad workaround to solve it: when the user selects the end date, add 1 day before saving in the DB. So the user would encode 2016-06-01 to 2016-06-01 but the DB would have 2016-06-01 to 2016-06-02... Which means always convert in both ways (from the DB to the interface, from the interface to the DB). Idk...

**Comment by arshaw** [comment]
@jnachtigall thanks for the extensive comment.

The ideal system you describe was actually existed in version 1.0 to 1.3. I changed it because many people found *that* system to be confusing. Devs were confused by the inconsistency between allDay dates being inclusive, and timed events being exclusive. I wish I could pull up all the past issues / emails regarding that, but would take a long time. It's really hard to please everybody.

I do understand what you are saying, but I agree more with the system that currently exists. Computer-understandable formats don't always need to be the same formats that humans read. For example, array indices start with `0`, but humans start counting at `1`!

Here's what I'd suggest... hopefully the *view* layer of your app is separate from the business logic. Store end dates with exclusive times (most apps do this). Whenever you are about to display an end time to the user, always subtract 1 day. Immediately after you accept an end date input from the user, always add 1 before you store it.

I know this is somewhat inconvenient, but the alternative has proven to be even more inconvenient to developers. I think the current approach is the lesser of two evils.

**Comment by jnachtigall** [comment]
@arshaw Thanks for your answer and also for the description on how to working around it. I see you've also altered the title if the issue (removing the 'option' part), but not explaining why just an option for this would not be possible? My suggestion is not about changing it globally but rather adding an option for inclusiveness instead (off by default).

**Comment by arshaw** [comment]
Ah ok, sorry for the missing the point. I'll change the title.

I'm hesitant to have a setting that alters such core behavior for two reasons:

1. will make the internal code more complicated. everywhere an end date is manipulated they'll have to be conditional behavior.
2. when people are using this behavior, and are viewing examples on the internet, the example might not match because and they'll get confused.

Eventually I want to move towards [Event Objects](http://fullcalendar.io/docs/event_data/Event_Object/) being prototype-based objects with real methods. When that happens, I can introduce two new convenience methods: `setNaturalEnd` and `getNaturalEnd`, which will do the +/- math for you. An event object could also accept a `naturalEnd` date input property.

What do you think?

**Comment by jnachtigall** [comment]
> Eventually I want to move towards Event Objects being prototype-based objects with real methods. When that happens, I can introduce two new convenience methods: setNaturalEnd and getNaturalEnd, which will do the +/- math for you. An event object could also accept a naturalEnd date input property.

Good idea, sounds like this would make working around easier. I am not too familar with fullcalendar internals, just used it for one project where I faced this problem. Would your proposal tackle the concerns as raised in https://github.com/fullcalendar/fullcalendar/issues/2344#issuecomment-132572231? I think that having an easy work around for this which is not too hacky and also safe concerning upgrades would be great.

I really like that it is very easy to get things done with fullcalendar. Hence, if the proposal would be amended by a short documentation at http://fullcalendar.io/docs/ like 'How to make the allDay end dates inclusive?' so that it's easy to find and google then this would be great. It's not enough to have the mentioned helper methods if one does not need how to use them to fullfil the goal.

**Comment by arshaw** [comment]
the people who don't like exclusive end dates, like those in #2344, could simply use the proposed "natural" getters/setters/properties and hopefully they would be happy.

i **definitely** need to improve the docs. made a ticket to do that:
https://github.com/fullcalendar/fullcalendar/issues/3205

if it's okay with you I will repurpose this ticket to be about the `naturalEnd` event object input property, as well as  `getNaturalEnd` / `setNaturalEnd` methods. this is dependent, on moving Event Objects over to prototype-based methods, which could take a while FYI.

Thanks for all the input.

**Comment by jhwinters** [comment]
Rather late I know, but I have a possibly helpful suggestion for atilleul.

You quite rightly say that your users don't want to have to enter 2016-06-01 to 2016-06-02 for a one day event - and they shouldn't have to.

The problem arises however because of the inconsistent way in which people refer to the duration of events.  When we say "From Mon 1st to Wed 3rd" we implicitly mean "From the start of Mon 1st to the end of Wed 3rd".  There is no reason (nor indeed feasibility) to get people to change, but it's important to identify the inconsistency in order to handle it in the cleanest way.

The point to apply the correction is at the point of data entry.  Thus when a user enters:

Start: 2016-06-01
End: 2016-06-03

meaning to create a 3 day event, your system should recognise what they mean and interpret appropriately.  The user means "from the start of Mon 1st to the end of Wed 3rd" and so that's what the system should store.  The start of Mon 1st is "2016-06-01 00:00:00" and the end of Wed 3rd is "2016-06-04 00:00:00"

Your data store (and not just FC) *needs* the information stored like this because otherwise it can't do correct calculations on duration or event overlap.  If you store the end date as 2016-06-03 in your database, you won't be able to do meaningful queries referencing that date field.  Comparing, say, "2016-06-03 09:00" to "2016-06-03" will indicate that the former is after the latter (which it is).  However "2016-06-03 09:00" is *not* after the end of your event.

Of course, when you come to show the timings in a textual form again (e.g. because the user is editing the event) you need to adjust the end time back to the user's expectations again.

Any other approach than this will inevitably lead to frig layered on frig layered on frig.  Fix the problem at the point of source (the fact that people use a slightly inconsistent notation) and then everything else flows smoothly.  You can't fix the people (nor should you want to) so fix the data at the point of entry.

Take a close look at a 3 day event displayed in FullCalendar and ask yourself where the coloured bar ends.

**Comment by birow** [comment]
I understand that there are those whose exclusive behavior is logical and some do not.

For me it would be consistent: when we accept the start date in allday mode for a full day, the closing date should also be accepted for a full day.

But referring to my first sentence my opinion is an option: allDayExclusive: true / false

**Comment by kirami** [comment]
I'm not sure I agree with the code.  On line 9238 of version fullcallendar.js v3.9.0  the line is:

if (endTimeMS && endTimeMS >= this.nextDayThreshold) 

if I set my next Day threshold to be 00:00:00 and my endtime is say June 3 00:00:00.  This is the first minute of June 3.  Even if I don't have allDay as true if my threshold is 0 this date/time should count as the next day.  When you evaluate if( 0 && 0>=0) it will obviously always fail.  I'm wondering if it's supposed to be if (endTimeMs != null && endTimeMS >= this.nextDayThreshold).  Just my opinion.

**Comment by jhwinters** [comment]
> if I set my next Day threshold to be 00:00:00 and my endtime is say June 3 00:00:00. This is the first minute of June 3.

I think possibly you're misinterpreting what 2019-06-03 00:00:00 is.  It's _not_ the first minute of the 3rd of June.  It's an instant (no duration) at the exact end of the 2nd of June and start of the 3rd.

The documentation for nextDayThreshold begins:

> When an event’s end time spans into another day,

If your event ends at 2019-06-03 00:00:00 then it does not span into another day, so isn't eligible for nextDayThreshold processing.

If you're dealing with all-day events (ending at exactly midnight) then the nextDayThreshold stuff isn't really a substitute for storing your dates correctly in the first place.  The issue of why you need to store the dates correctly has been done to death in this thread and others linked from it, so I won't repeat the explanation.

Suffice to say that if you're trying to fix it with nextDayThreshold then you're addressing the symptoms, rather than fixing your original data storage error.
